### PR TITLE
resolved-ts: use min_lock_ts-1 as the candidate of resolved-ts

### LINKED ISSUE #17728: Resolved-ts must be strictly smaller than locks' min_commit_ts when calculating it
## Bug Report

ref #17459

By the definition of resolved-ts, we require:
 `commit_ts > resolved_ts`, if the commit happens after the calculation of the resolved-ts.

Our implementation uses min_commit_ts to calculate resolved-ts, so that what we have is: 
`commit_ts >= min_commit_ts >= resolved_ts`

`resolved_ts` may be equal to `commit_ts`, which breaks the definition of resolved-ts.

NOTE: This bug **cannot happen in current(v8.4) implementation**. 
- Only Pipelined DMLs use min_commit_ts for resolved-ts calculation
- These min_commit_ts used are (indirectly) from PD TSO
- This ensures their commit_ts is always greater than min_commit_ts
- Therefore, commit_ts is guaranteed to be greater than resolved-ts

### PR DESCRIPTION
**Pull request by ekexium**
### What is changed and how it works?

Issue Number: Close #17728

What's Changed:

#### The problem
By the definition of resolved-ts, we require:
 `commit_ts > resolved_ts`, if the commit happens after the calculation of the resolved-ts.

Our implementation uses min_commit_ts to calculate resolved-ts, so that what we have is: 
`commit_ts >= min_commit_ts >= resolved_ts`

`resolved_ts` may be equal to `commit_ts`, which breaks the definition of resolved-ts.

#### The fix
This PR ensures `min_commit_ts > resolved_ts` to fix the problem

```commit-message
Use min_lock_ts-1 as the candidate of resolved-ts, to ensure resolved_ts < lock.min_commit_ts( <= commit_ts).
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/resolved_ts/src/resolver.rs [hunk 1]
```diff
@@ -432,7 +432,12 @@ impl Resolver {
         // Find the min start ts.
         let min_lock = self.oldest_transaction();
         let has_lock = min_lock.is_some();
-        let min_txn_ts = min_lock.as_ref().map(|(ts, _)| *ts).unwrap_or(min_ts);
+        // min_txn_ts is min_lock.ts - 1, because we need to guarantee
```
Comment (cfzjywxk) [inline components/resolved_ts/src/resolver.rs:435]: Would it be a problem if all value of the `min_lock.ts` comes from `start_ts` of transactions, and `min(all_transaction_start_ts) - 1` is used to calculate which differs from before?

Comment (ekexium) [reply components/resolved_ts/src/resolver.rs:435]: I suppose the resolved-ts can be arbitrarily smaller. Setting a lower resolved-ts imposes a stricter constraint on resolved-ts users. The main inconvenience I can think of is attempting to map a resolved-ts to its corresponding transaction.

Comment (ekexium) [reply components/resolved_ts/src/resolver.rs:435]: The resolved-ts may decrease by 1 during TiKV upgrades. This might be a problem.

**Review by cfzjywxk — APPROVED** [review]
LGTM

File: components/resolved_ts/src/resolver.rs [hunk 2]
```diff
@@ -568,8 +568,17 @@ impl Resolver {
         self.read_progress.as_ref()
     }
 
-    // Return the transaction with the smallest min_commit_ts. When min_commit_ts
-    // is unknown, use start_ts instead.
+    // Returns the minimum possible (commit_ts - 1), based on the knowledge we have,
+    // i.e. from all locks currently being tracked.
+    // The function is to provide an upper bound of resolved-ts. By definition
+    // resolved-ts must be strictly larger than a future commit_ts.
```
Comment (you06) [inline components/resolved_ts/src/resolver.rs:574]:
```suggestion
    // resolved-ts must be strictly smaller than a future commit_ts.
```

If I understand correctly, it's `resolved-ts < commit-ts`.

**Review by you06 — APPROVED** [review]

**Pull request merged**